### PR TITLE
Add Attenuation Model "DISABLED" for AudioStreamPlayer3D

### PR DESCRIPTION
**Pull request by NilsIrl**
Pull request addressing #26538 
I decided to do the pull request cause I had to patch Godot for my own work anyway.

### CODE REVIEW

File: doc/classes/AudioStreamPlayer3D.xml
Comment (YeldhamDev) [inline doc/classes/AudioStreamPlayer3D.xml]: Forgot to edit the name here.

**Comment by YeldhamDev** [comment]
It would be better if it was named `NONE` instead of just `NO`.

File: scene/3d/audio_stream_player_3d.cpp
Comment (YeldhamDev) [inline scene/3d/audio_stream_player_3d.cpp]: Incorrect indentation.

**Comment by YeldhamDev** [comment]
Remember to [squash your commits](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase).

File: doc/classes/AudioStreamPlayer3D.xml
Comment (YeldhamDev) [inline doc/classes/AudioStreamPlayer3D.xml]: One last thing: a [comma](https://en.wikipedia.org/wiki/Serial_comma) after "logarithmically".

**Comment by NilsIrl** [comment]
How can I reproduce the Travis CI builds because I tried using the settings that failed `platform=x11 tools=no target=release cache-name=${PLATFORM}-clang extra_args="werror=yes" use_llvm=yes` and my local build doesn't fail.

**Comment by YeldhamDev** [comment]
I also don't know why it's failing.
@akien-mga Could you shed a little light here?

**Comment by akien-mga** [comment]
It's in the logs: 
```
scene/3d/audio_stream_player_3d.cpp:835:2: error: result of comparison of
      constant 4 with expression of type 'AudioStreamPlayer3D::AttenuationModel'
      is always false [-Werror,-Wtautological-constant-out-of-range-compare]
        ERR_FAIL_INDEX(p_model, 4);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~
./core/error_macros.h:136:43: note: expanded from macro 'ERR_FAIL_INDEX'
                if (unlikely((m_index) < 0 || (m_index) >= (m_size))) ...
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
./core/typedefs.h:317:41: note: expanded from macro 'unlikely'
#define unlikely(x) __builtin_expect(!!(x), 0)
                                        ^
1 error generated.
```

**Comment by YeldhamDev** [comment]
@akien-mga I saw that, but since a new one was added, shouldn't that be correct?

**Comment by NilsIrl** [comment]
> @akien-mga I saw that, but since a new one was added, shouldn't that be correct?

I don't understand either

File: scene/3d/audio_stream_player_3d.cpp
Comment (NilsIrl) [inline scene/3d/audio_stream_player_3d.cpp]: When I set the value to three, in the editor, None appears but I cannot select it. When the value is set to 4, then the build fails (gives warning).

Also it says value is always false, but shouldn't that be the case?

Comment (NilsIrl) [inline scene/3d/audio_stream_player_3d.cpp]: When I set the value to three, in the editor, None appears but I cannot select it. When the value is set to 4, then the build fails (gives warning).

Also it says value is always false, but shouldn't that be the case?

Comment (bojidar-bg) [reply scene/3d/audio_stream_player_3d.cpp]: Well, maybe just remove the ERR_FAIL_INDEX. It is just saying that something of type AttenuationModel, which is an enum of 4 values, can never be more than or equal to 4.

Comment (akien-mga) [reply scene/3d/audio_stream_player_3d.cpp]: The problem is that this check is useless for the enumeration value, but your can pass an arbitrary int from GDScript. So we still need to check that we're in bounds, no?

For the reference I've seen these warnings a while ago too, but it's weird that they get raised in this PR when the exact same warning was valid before too.

Comment (NilsIrl) [reply scene/3d/audio_stream_player_3d.cpp]: I don't know if it is really important but here are the errors raised when set to 3
```
   At: scene/3d/audio_stream_player_3d.cpp:835.
ERROR: set_attenuation_model: Index p_model=3 out of size (3=3)
   At: scene/3d/audio_stream_player_3d.cpp:835.
ERROR: set_attenuation_model: Index p_model=3 out of size (3=3)
   At: scene/3d/audio_stream_player_3d.cpp:835.
ERROR: set_attenuation_model: Index p_model=3 out of size (3=3)
```

File: scene/3d/audio_stream_player_3d.h
Comment (reduz) [inline scene/3d/audio_stream_player_3d.h]: I would make it ATTENUATION_DISABLED to be more in line with the rest of Godot

File: scene/3d/audio_stream_player_3d.cpp
Comment (reduz) [inline scene/3d/audio_stream_player_3d.cpp]: Would just make it "Disabled"

**Review by reduz — CHANGES_REQUESTED** [review]

File: scene/3d/audio_stream_player_3d.cpp
Comment (NilsIrl) [reply scene/3d/audio_stream_player_3d.cpp]: I changed it to disabled.

**Comment by NilsIrl** [comment]
We still have the build failing

**Comment by reduz** [comment]
@akien-mga This should be good to merge once clang passes I guess

**Review by reduz — APPROVED** [review]

File: scene/3d/audio_stream_player_3d.cpp
Comment (starry-abyss) [reply scene/3d/audio_stream_player_3d.cpp]: I suspect if add a cast, then the warning should go:
`if (unlikely((m_index) < 0 || ((int) m_index) >= (m_size)))`

Probably this particular compiler (version) is smart enough to figure out what Bojidar said, and since warnings are forced as errors, it can't compile

**Comment by NilsIrl** [comment]
> https://bugs.llvm.org//show_bug.cgi?id=16154
This seems to be somewhat of a known issue. However it doesn't explain why the tests pass when the constant is set to 3.

**Comment by starry-abyss** [comment]
I don't think it's a compiler bug. IMHO it's just Godot code abuses the type system and doesn't add a cast stating to compiler that the abuse was intentional.

> However it doesn't explain why the tests pass when the constant is set to 3.
Because 3 has a corresponding enum value, so other enum value can give either a) less than 3 or b) equal to 3.
And 4 doesn't have an enum value, so it's always a) case, which the compiler hints at with the warning.
Apparently, we want this check only at runtime (when receiving values from GDScript), but the compiler does also the same at compile time based on types given.

File: scene/3d/audio_stream_player_3d.cpp
Comment (NilsIrl) [reply scene/3d/audio_stream_player_3d.cpp]: Sorry I didn't notice your comment.
Would it be better to cast at the macro call like so:
`ERR_FAIL_INDEX((int) p_model, 4);`
or like you proposed in the macro?

Comment (starry-abyss) [reply scene/3d/audio_stream_player_3d.cpp]: Well, it's not up to me to decide.
I think adding it in (potentially) every call of this macro is too much hassle.
And I don't think casting here breaks any stuff, because for comparison with a number we are implicitly casting the enum value anyway. Though it might belong in a separate PR :)
As for signed/unsigned, generally signed is better by default, as it helps avoid sneaky bugs with conversion between signed and unsigned.

**Comment by starry-abyss** [comment]
From what I read, now it compiles OK, but complains about style, sorry, my bad. It seems to want the line to look like:
`if (unlikely((m_index) < 0 || ((int)m_index) >= (m_size))) {`

so there is no space before `m_index`.

File: core/error_macros.h
Comment (akien-mga) [inline core/error_macros.h]: Seems good to me, but it should likely be done in a separate PR and also in `ERR_FAIL_INDEX_V`.
Existing casts to `int` could also be removed:
```
$ rg ERR_FAIL_INDEX.*\\\(\\\(int\\\)
scene/resources/animation.cpp
1917:   ERR_FAIL_INDEX((int)p_mode, 4);

scene/2d/camera_2d.cpp
452:    ERR_FAIL_INDEX((int)p_margin, 4);
459:    ERR_FAIL_INDEX_V((int)p_margin, 4, 0);
476:    ERR_FAIL_INDEX((int)p_margin, 4);
483:    ERR_FAIL_INDEX_V((int)p_margin, 4, 0);

scene/gui/texture_progress.cpp
62:     ERR_FAIL_INDEX((int)p_margin, 4);
69:     ERR_FAIL_INDEX_V((int)p_margin, 4, 0);

drivers/gles3/rasterizer_scene_gles3.cpp
4720:           ERR_FAIL_INDEX((int)shadow, shadow_atlas->quadrants[quadrant].shadows.size());

scene/gui/control.cpp
2251:   ERR_FAIL_INDEX((int)p_margin, 4);
2257:   ERR_FAIL_INDEX_V((int)p_margin, 4, NodePath());

scene/3d/audio_stream_player_3d.cpp
844:    ERR_FAIL_INDEX((int)p_mode, 2);

scene/resources/style_box.cpp
146:    ERR_FAIL_INDEX((int)p_margin, 4);
210:    ERR_FAIL_INDEX((int)p_expand_margin, 4);
233:    ERR_FAIL_INDEX_V((int)p_expand_margin, 4, 0);

drivers/gles2/rasterizer_scene_gles2.cpp
3006:           ERR_FAIL_INDEX((int)shadow, shadow_atlas->quadrants[quadrant].shadows.size());

scene/2d/collision_polygon_2d.cpp
218:    ERR_FAIL_INDEX((int)p_mode, 2);

scene/gui/label.cpp
513:    ERR_FAIL_INDEX((int)p_align, 4);
525:    ERR_FAIL_INDEX((int)p_align, 4);

scene/gui/line_edit.cpp
533:    ERR_FAIL_INDEX((int)p_align, 4);

scene/gui/nine_patch_rect.cpp
113:    ERR_FAIL_INDEX((int)p_margin, 4);
135:    ERR_FAIL_INDEX_V((int)p_margin, 4, 0);

core/os/input.cpp
45:     ERR_FAIL_INDEX((int)p_mode, 4);
```

Comment (akien-mga) [reply core/error_macros.h]: And same for `CRASH_BAD_INDEX` I guess.

Comment (akien-mga) [reply core/error_macros.h]: Alternatively you could just do the `(int)` cast in this PR, since it's already done in various other places.
We can then do and review the `error_macros.h` change in another PR, without blocking this one.

Comment (starry-abyss) [reply core/error_macros.h]: And for ERR_FAIL_UNSIGNED_INDEX_V with cast to (unsigned int) I think...
OK, I'll do the PR then, unless somebody is working on it already.

Comment (NilsIrl) [reply core/error_macros.h]: So in the current commit e9ac737fe, the cast is done at the macro call but with the intent of changing that later in another PR.

Comment (akien-mga) [reply core/error_macros.h]: Perfect, thanks!

Comment (starry-abyss) [reply core/error_macros.h]: @NilsIrl I'm on it!

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
